src/core/thread/meshcop_dataset_manager.cpp

```diff
@@ -356,6 +359,47 @@ ThreadError DatasetManager::Set(Coap::Header &aHeader, Message &aMessage, const
     mNetworkDataLeader.IncrementVersion();
     mNetworkDataLeader.IncrementStableVersion();
 
+    // notify commissioner if update is from thread device
+    if (!isUpdateFromCommissioner)
+    {
+        uint8_t *cur;
+        uint8_t *end;
+        uint8_t length;
+        uint16_t locator = 0;
+        Ip6::Address destination;
+
+        cur = mNetworkDataLeader.GetCommissioningData(length);
+        end = cur + length;
+
+        while (cur < end)
+        {
+            Tlv *data = reinterpret_cast<Tlv *>(cur);
+
+            if (data->GetType() == Tlv::kBorderAgentLocator)
+            {
+                locator = static_cast<BorderAgentLocatorTlv *>(data)->GetBorderAgentLocator();
+            }
+
+            cur += sizeof(Tlv) + data->GetLength();
+        }
+
+        if (locator != 0)
+        {
+            memset(&destination, 0, sizeof(destination));
+            memcpy(&destination, mNetif.GetMle().GetMeshLocal16(), 8);
+            destination.mFields.m16[4] = HostSwap16(0x0000);
+            destination.mFields.m16[5] = HostSwap16(0x00ff);
+            destination.mFields.m16[6] = HostSwap16(0xfe00);
+            destination.mFields.m16[7] = HostSwap16(locator);
+        }
+        else
+        {
+            mNetif.GetCommissioner().GetCommissionerAddress(destination);
```

Comment:
If an on-mesh commissioner is active, it must have petitioned.  The leader will add the Border Agent Locator TLV to the Commissioner Dataset when it accepts a new Commissioner.